README.md
# nativescript-angular-jwt

[![npm version](https://badge.fury.io/js/%40auth0%2Fangular-jwt.svg)](https://badge.fury.io/js/%40auth0%2Fangular-jwt)

### **NOTE:** This library is now at version 5 and is published on npm as `nativescript-angular-jwt`. If you're looking for the pre-v1.0 version of this library, it can be found in the `pre-v1.0` branch and on npm as `angular2-jwt`.

**Version v5 of this library has some breaking changes concerning the `allowedDomains` and `disallowedRoutes`.**

**nativescript-angular-jwt v5 is to be used with Angular v10+ and RxJS v6+. For Angular v6+ to v9, use nativescript-angular-jwt v4**

This library provides an `HttpInterceptor` which automatically attaches a [JSON Web Token](https://jwt.io) to `HttpClient` requests.

This library does not have any functionality for (or opinion about) implementing user authentication and retrieving JWTs to begin with. Those details will vary depending on your setup, but in most cases, you will use a regular HTTP request to authenticate your users and then save their JWTs in local storage or in a cookie if successful.

> **Note:** This library can only be used with Angular 4.3 and higher because it relies on an `HttpInterceptor` from Angular's `HttpClient`. This feature is not available on lower versions.

## Installation

```bash
# installation with npm
npm install nativescript-angular-jwt

# installation with yarn
yarn add nativescript-angular-jwt
```

## Usage: Standalone

If you are only interested in the JWT Decoder, and are not interested in extended
injectable features, you can simply create an instance of the utility and use it
directly:

```ts
import { JwtHelperService } from "nativescript-angular-jwt";

const helper = new JwtHelperService();

const decodedToken = helper.decodeToken(myRawToken);
const expirationDate = helper.getTokenExpirationDate(myRawToken);
const isExpired = helper.isTokenExpired(myRawToken);
```

## Usage: Injection

Import the `JwtModule` module and add it to your imports list. Call the `forRoot` method and provide a `tokenGetter` function. You must also add any domains to the `allowedDomains`, that you want to make requests to by specifying an `allowedDomains` array.

Be sure to import the `HttpClientModule` as well.

```ts
import { JwtModule } from "nativescript-angular-jwt";
import { HttpClientModule } from "@angular/common/http";

export function tokenGetter() {
  return localStorage.getItem("access_token");
}

@NgModule({
  bootstrap: [AppComponent],
  imports: [
    // ...
    HttpClientModule,
    JwtModule.forRoot({
      config: {
        tokenGetter: tokenGetter,
        allowedDomains: ["example.com"],
        disallowedRoutes: ["http://example.com/examplebadroute/"],
      },
    }),
  ],
})
export class AppModule {}
```

Any requests sent using Angular's `HttpClient` will automatically have a token attached as an `Authorization` header.

```ts
import { HttpClient } from "@angular/common/http";

export class AppComponent {
  constructor(public http: HttpClient) {}

  ping() {
    this.http.get("http://example.com/api/things").subscribe(
      (data) => console.log(data),
      (err) => console.log(err)
    );
  }
}
```

## Configuration Options

### `tokenGetter: function(HttpRequest): string`

The `tokenGetter` is a function which returns the user's token. This function simply needs to make a retrieval call to wherever the token is stored. In many cases, the token will be stored in local storage or session storage.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    tokenGetter: () => {
      return localStorage.getItem("access_token");
    },
  },
});
```

If you have multiple tokens for multiple domains, you can use the `HttpRequest` passed to the `tokenGetter` function to get the correct token for each intercepted request.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    tokenGetter: (request) => {
      if (request.url.includes("foo")) {
        return localStorage.getItem("access_token_foo");
      }

      return localStorage.getItem("access_token");
    },
  },
});
```

### `allowedDomains: array`

Authenticated requests should only be sent to domains you know and trust. Many applications make requests to APIs from multiple domains, some of which are not controlled by the developer. Since there is no way to know what the API being called will do with the information contained in the request, it is best to not send the user's token to all APIs in a blind fashion.

List any domains you wish to allow authenticated requests to be sent to by specifying them in the `allowedDomains` array. **Note that standard http port 80 and https port 443 requests don't require a port to be specified. A port is only required in the allowed domains host name if you are authenticating against a non-standard port e.g. localhost:3001**

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    allowedDomains: ["localhost:3001", "foo.com", "bar.com"],
  },
});
```

### `disallowedRoutes: array`

If you do not want to replace the authorization headers for specific routes, list them here. This can be useful if your
initial auth route(s) are on an allowed domain and take basic auth headers. These routes need to be prefixed with the correct protocol (`http://`, `https://`). If you want to add a route to the list of disallowed routes regardless of the protocol, you can prefix it with `//`.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    disallowedRoutes: [
      "http://localhost:3001/auth/",
      "https://foo.com/bar/",
      "//foo.com/bar/baz",
      /localhost:3001\/foo\/far.*/,
    ], // strings and regular expressions
  },
});
```

**Note:** If requests are sent to the same domain that is serving your Angular application, you do not need to add that domain to the `allowedDomains` array. However, this is only the case if you don't specify the domain in the `Http` request.

For example, the following request assumes that the domain is the same as the one serving your app. It doesn't need to be allowed in this case.

```ts
this.http.get('/api/things')
  .subscribe(...)
```

However, if you are serving your API at the same domain as that which is serving your Angular app **and** you are specifying that domain in `Http` requests, then it **does** need to be explicitely allowed.

```ts
// Both the Angular app and the API are served at
// localhost:4200 but because that domain is specified
// in the request, it must be allowed
this.http.get('http://localhost:4200/api/things')
  .subscribe(...)
```

### `headerName: string`

The default header name is `Authorization`. This can be changed by specifying a custom `headerName` which is to be a string value.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    headerName: "Your Header Name",
  },
});
```

### `authScheme: string | function(HttpRequest): string`

The default authorization scheme is `Bearer` followed by a single space. This can be changed by specifying a custom `authScheme`. You can pass a string which will prefix the token for each request.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    authScheme: "Basic ",
  },
});
```

If you want to change the auth scheme dynamically, or based on the request, you can configure a getter function which returns a string.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    authScheme: (request) => {
      if (request.url.includes("foo")) {
        return "Basic ";
      }

      return "Bearer ";
    },
  },
});
```

### `throwNoTokenError: boolean`

Setting `throwNoTokenError` to `true` will result in an error being thrown if a token cannot be retrieved with the `tokenGetter` function. Defaults to `false`.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    throwNoTokenError: true,
  },
});
```

### `skipWhenExpired: boolean`

By default, the user's JWT will be sent in `HttpClient` requests even if it is expired. You may choose to not allow the token to be sent if it is expired by setting `skipWhenExpired` to true.

```ts
// ...
JwtModule.forRoot({
  config: {
    // ...
    skipWhenExpired: true,
  },
});
```

## Using a Custom Options Factory Function

In some cases, you may need to provide a custom factory function to properly handle your configuration options. This is the case if your `tokenGetter` function relies on a service or if you are using an asynchronous storage mechanism (like Ionic's `Storage`).

Import the `JWT_OPTIONS` `InjectionToken` so that you can instruct it to use your custom factory function.

Create a factory function and specify the options as you normally would if you were using `JwtModule.forRoot` directly. If you need to use a service in the function, list it as a parameter in the function and pass it in the `deps` array when you provide the function.

```ts
import { JwtModule, JWT_OPTIONS } from 'nativescript-angular-jwt';
import { TokenService } from './app.tokenservice';

// ...

export function jwtOptionsFactory(tokenService) {
  return {
    tokenGetter: () => {
      return tokenService.getAsyncToken();
    },
    allowedDomains: ["example.com"]
  }
}

// ...

@NgModule({
  // ...
  imports: [
    JwtModule.forRoot({
      jwtOptionsProvider: {
        provide: JWT_OPTIONS,
        useFactory: jwtOptionsFactory,
        deps: [TokenService]
      }
    })
  ],
  providers: [TokenService]
})
```

**Note:**: If a `jwtOptionsFactory` is defined, then `config` is ignored. _Both configuration alternatives can't be defined at the same time_.

## Configuration for Ionic 2+

The custom factory function approach described above can be used to get a token asynchronously with Ionic's `Storage`.

```ts
import { JwtModule, JWT_OPTIONS } from 'nativescript-angular-jwt';
import { Storage } from '@ionic/storage';

export function jwtOptionsFactory(storage) {
  return {
    tokenGetter: () => {
      return storage.get('access_token');
    },
    allowedDomains: ["example.com"]
  }
}

// ...

@NgModule({
  // ...
  imports: [
    JwtModule.forRoot({
      jwtOptionsProvider: {
        provide: JWT_OPTIONS,
        useFactory: jwtOptionsFactory,
        deps: [Storage]
      }
    })
  ]
})
```

**Note:**: If a `jwtOptionsFactory` is defined, then `config` is ignored. _Both configuration alternatives can't be defined at the same time_.

## Configuration Options

### `JwtHelperService: service`

This service contains helper functions:

## isTokenExpired (old tokenNotExpired function)

```
import { JwtHelperService } from 'nativescript-angular-jwt';
// ...
constructor(public jwtHelper: JwtHelperService) {}

ngOnInit() {
console.log(this.jwtHelper.isTokenExpired()); // true or false
}
```

## getTokenExpirationDate

```
import { JwtHelperService } from 'nativescript-angular-jwt';
// ...
constructor(public jwtHelper: JwtHelperService) {}

ngOnInit() {
console.log(this.jwtHelper.getTokenExpirationDate()); // date
}
```

## decodeToken

```
import { JwtHelperService } from 'nativescript-angular-jwt';
// ...
constructor(public jwtHelper: JwtHelperService) {}

ngOnInit() {
console.log(this.jwtHelper.decodeToken(token)); // token
}
```

## What is Auth0?

Auth0 helps you to:

- Add authentication with [multiple authentication sources](https://auth0.com/docs/identityproviders), either social like **Google, Facebook, Microsoft Account, LinkedIn, GitHub, Twitter, Box, Salesforce, among others**, or enterprise identity systems like **Windows Azure AD, Google Apps, Active Directory, ADFS or any SAML Identity Provider**.
- Add authentication through more traditional **[username/password databases](https://auth0.com/docs/connections/database/custom-db)**.
- Add support for **[linking different user accounts](https://auth0.com/docs/link-accounts)** with the same user.
- Support for generating signed [Json Web Tokens](https://auth0.com/docs/jwt) to call your APIs and **flow the user identity** securely.
- Analytics of how, when and where users are logging in.
- Pull data from other sources and add it to the user profile, through [JavaScript rules](https://auth0.com/docs/rules/current).

## Create a free Auth0 account

1.  Go to [Auth0](https://auth0.com/signup) and click Sign Up.
2.  Use Google, GitHub or Microsoft Account to login.

## Issue Reporting

If you have found a bug or if you have a feature request, please report them at this repository issues section. Please do not report security vulnerabilities on the public GitHub issue tracker. The [Responsible Disclosure Program](https://auth0.com/whitehat) details the procedure for disclosing security issues.

## Author

[Auth0](auth0.com)

## License

This project is licensed under the MIT license. See the [LICENSE](LICENSE.txt) file for more info.


CHANGELOG.md
# Change log

## Version [5.0.0](https://github.com/nativescript-angular-jwt/tags/v5.0.0)

**Warning: this version has some breaking changes concerning the allowed domains and dissalowed routes!**

- Replace whitelistedDomains to allowedDomain[#668](https://github.com/nativescript-angular-jwt/pull/668)
- Replace blacklistedRoutes to disallowedRoutes[#668](https://github.com/nativescript-angular-jwt/pull/668)
- Removed the url dependency, as this is a Node module in the CommonJS format, and the Angular 10 CLI throws warnings when using dependencies in the CommonJS format. We're using the default URL interface, https://developer.mozilla.org/en-US/docs/Web/API/URL [#666](https://github.com/nativescript-angular-jwt/pull/666)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/v4.2.0..v5.0.0)

## Version [4.2.0](https://github.com/nativescript-angular-jwt/tags/v4.2.0)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/v4.1.2..v4.2.0)

- Allow the authScheme config parameter to be a getter function [#659](https://github.com/nativescript-angular-jwt/pull/659)

## Version [4.1.2](https://github.com/nativescript-angular-jwt/tags/v4.1.2) (2020-05-16)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/v4.1.1..v4.1.2)

- Support domains with a port other than the default HTTP ports (HTTP: 80, HTTPS: 443)
  [#656](https://github.com/nativescript-angular-jwt/pull/656)

## Version [4.1.1](https://github.com/nativescript-angular-jwt/tags/v4.1.1) (2020-05-15)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/v4.1.0..v4.1.1)

- Something went wrong pulishing `v4.1.0`, this version fixes that.

## Version [4.1.0](https://github.com/nativescript-angular-jwt/tags/v4.1.0) (2020-05-15)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/4.0.0..v4.1.0)

- Use blacklist domains regardless of their protocol [#644](https://github.com/nativescript-angular-jwt/pull/644)
- Pass the HttpRequest to the tokenGetter [#649](https://github.com/nativescript-angular-jwt/pull/649)

## Version [4.0.0](https://github.com/nativescript-angular-jwt/tags/v4.0.0) (2020-02-07)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/3.0.1..4.0.0)

From [\#622](https://github.com/nativescript-angular-jwt/pull/622) [avatsaev](https://github.com/avatsaev):

- Angular 9 compatibility
- Angular Ivy compatibility

## Version [3.0.1](https://github.com/nativescript-angular-jwt/tags/v3.0.1) (2019-10-28)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/3.0.0..3.0.1)

- Be sure to handle `undefined` tokens [\#626](https://github.com/nativescript-angular-jwt/pull/626) [@RobertRad](https://github.com/RobertRad)
- docs: corrected/rephrased a sentence in README.md [\#625](https://github.com/nativescript-angular-jwt/pull/625) [@noopur-tiwari](https://github.com/noopur-tiwari)

## Version [3.0.0](https://github.com/nativescript-angular-jwt/releases/tag/3.0.0) (2019-07-16)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/2.1.2..3.0.0)

- **Breaking change** `isTokenExpired` now returns `false` if no expiry date is found inside the token. This is a change to align with the [JWT spec](https://tools.ietf.org/html/rfc7519#section-4.1.4), but may break applications that rely on the previous behavior. [#562](https://github.com/nativescript-angular-jwt/pull/562) [@atom-morgan](https://github.com/atom-morgan)

## Version [2.1.2](https://github.com/nativescript-angular-jwt/releases/tag/2.1.2) (2019-07-15)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/2.1.1..2.1.2)

- Gracefully handle null/empty tokens [#586](https://github.com/nativescript-angular-jwt/pull/586)

## Version [2.1.1](https://github.com/nativescript-angular-jwt/releases/tag/2.1.1) (2019-07-01)

[Full Changelog](https://github.com/nativescript-angular-jwt/compare/2.1.0...2.1.1)

- Blacklist/Whitelist check fix [#538](https://github.com/nativescript-angular-jwt/pull/538)
- Refactor deep rxjs imports and use named define [#608](https://github.com/nativescript-angular-jwt/pull/608)
- fix(rxjs): remove imports from rxjs/internal [#542](https://github.com/nativescript-angular-jwt/pull/542)

> Note: historical changelog information has not been recorded in this format. Please see the [releases page](https://github.com/nativescript-angular-jwt/releases) for information on previous releases.


bundles/nativescript-angular-jwt.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('rxjs/operators'), require('rxjs'), require('@angular/common/http')) :
    typeof define === 'function' && define.amd ? define('nativescript-angular-jwt', ['exports', '@angular/core', '@angular/common', 'rxjs/operators', 'rxjs', '@angular/common/http'], factory) :
    (global = global || self, factory((global.auth0 = global.auth0 || {}, global.auth0['nativescript-angular-jwt'] = {}), global.ng.core, global.ng.common, global.rxjs.operators, global.rxjs, global.ng.common.http));
}(this, (function (exports, core, common, operators, rxjs, http) { 'use strict';

    var JWT_OPTIONS = new core.InjectionToken('JWT_OPTIONS');

    // tslint:disable:no-bitwise
    var JwtHelperService = /** @class */ (function () {
        function JwtHelperService(config) {
            if (config === void 0) { config = null; }
            this.tokenGetter = (config && config.tokenGetter) || function () { };
        }
        JwtHelperService.prototype.urlBase64Decode = function (str) {
            var output = str.replace(/-/g, "+").replace(/_/g, "/");
            switch (output.length % 4) {
                case 0: {
                    break;
                }
                case 2: {
                    output += "==";
                    break;
                }
                case 3: {
                    output += "=";
                    break;
                }
                default: {
                    throw new Error("Illegal base64url string!");
                }
            }
            return this.b64DecodeUnicode(output);
        };
        // credits for decoder goes to https://github.com/atk
        JwtHelperService.prototype.b64decode = function (str) {
            var chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
            var output = "";
            str = String(str).replace(/=+$/, "");
            if (str.length % 4 === 1) {
                throw new Error("'atob' failed: The string to be decoded is not correctly encoded.");
            }
            for (
            // initialize result and counters
            var bc = 0, bs = void 0, buffer = void 0, idx = 0; 
            // get next character
            (buffer = str.charAt(idx++)); 
            // character found in table? initialize bit storage and add its ascii value;
            ~buffer &&
                ((bs = bc % 4 ? bs * 64 + buffer : buffer),
                    // and if not first of each 4 characters,
                    // convert the first 8 bits to one ascii character
                    bc++ % 4)
                ? (output += String.fromCharCode(255 & (bs >> ((-2 * bc) & 6))))
                : 0) {
                // try to find character in table (0-63, not found => -1)
                buffer = chars.indexOf(buffer);
            }
            return output;
        };
        JwtHelperService.prototype.b64DecodeUnicode = function (str) {
            return decodeURIComponent(Array.prototype.map
                .call(this.b64decode(str), function (c) {
                return "%" + ("00" + c.charCodeAt(0).toString(16)).slice(-2);
            })
                .join(""));
        };
        JwtHelperService.prototype.decodeToken = function (token) {
            if (token === void 0) { token = this.tokenGetter(); }
            if (!token || token === "") {
                return null;
            }
            var parts = token.split(".");
            if (parts.length !== 3) {
                throw new Error("The inspected token doesn't appear to be a JWT. Check to make sure it has three parts and see https://jwt.io for more.");
            }
            var decoded = this.urlBase64Decode(parts[1]);
            if (!decoded) {
                throw new Error("Cannot decode the token.");
            }
            return JSON.parse(decoded);
        };
        JwtHelperService.prototype.getTokenExpirationDate = function (token) {
            if (token === void 0) { token = this.tokenGetter(); }
            var decoded;
            decoded = this.decodeToken(token);
            if (!decoded || !decoded.hasOwnProperty("exp")) {
                return null;
            }
            var date = new Date(0);
            date.setUTCSeconds(decoded.exp);
            return date;
        };
        JwtHelperService.prototype.isTokenExpired = function (token, offsetSeconds) {
            if (token === void 0) { token = this.tokenGetter(); }
            if (!token || token === "") {
                return true;
            }
            var date = this.getTokenExpirationDate(token);
            offsetSeconds = offsetSeconds || 0;
            if (date === null) {
                return false;
            }
            return !(date.valueOf() > new Date().valueOf() + offsetSeconds * 1000);
        };
        JwtHelperService.prototype.getAuthScheme = function (authScheme, request) {
            if (typeof authScheme === "function") {
                return authScheme(request);
            }
            return authScheme;
        };
        return JwtHelperService;
    }());
    JwtHelperService.decorators = [
        { type: core.Injectable }
    ];
    JwtHelperService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [JWT_OPTIONS,] }] }
    ]; };

    var JwtInterceptor = /** @class */ (function () {
        function JwtInterceptor(config, jwtHelper, document) {
            this.jwtHelper = jwtHelper;
            this.document = document;
            this.standardPorts = ["80", "443"];
            this.tokenGetter = config.tokenGetter;
            this.headerName = config.headerName || "Authorization";
            this.authScheme =
                config.authScheme || config.authScheme === ""
                    ? config.authScheme
                    : "Bearer ";
            this.allowedDomains = config.allowedDomains || [];
            this.disallowedRoutes = config.disallowedRoutes || [];
            this.throwNoTokenError = config.throwNoTokenError || false;
            this.skipWhenExpired = config.skipWhenExpired;
        }
        JwtInterceptor.prototype.isAllowedDomain = function (request) {
            if(!this.document || !this.document.location) return true;

            var origin = this.document && this.document.location ? this.document.location.origin : null;
            var requestUrl = new URL(request.url, origin);
            // If the host equals the current window origin,
            // the domain is allowed by default
            if (origin && (requestUrl.host === this.document.location.host)) {
                return true;
            }
            // If not the current domain, check the allowed list
            var hostName = "" + requestUrl.hostname + (requestUrl.port && !this.standardPorts.includes(requestUrl.port)
                ? ":" + requestUrl.port
                : "");
            return (this.allowedDomains.findIndex(function (domain) { return typeof domain === "string"
                ? domain === hostName
                : domain instanceof RegExp
                    ? domain.test(hostName)
                    : false; }) > -1);
        };
        JwtInterceptor.prototype.isDisallowedRoute = function (request) {
            var _this = this;
            if(!this.document || !this.document.location) return false;
            var origin = this.document && this.document.location ? this.document.location.origin : null;
            var requestedUrl = new URL(request.url, origin);
            return (this.disallowedRoutes.findIndex(function (route) {
                if (typeof route === "string") {
                    var parsedRoute = new URL(route, origin);
                    return (parsedRoute.hostname === requestedUrl.hostname &&
                        parsedRoute.pathname === requestedUrl.pathname);
                }
                if (route instanceof RegExp) {
                    return route.test(request.url);
                }
                return false;
            }) > -1);
        };
        JwtInterceptor.prototype.handleInterception = function (token, request, next) {
            var _a;
            var authScheme = this.jwtHelper.getAuthScheme(this.authScheme, request);
            var tokenIsExpired = false;
            if (!token && this.throwNoTokenError) {
                throw new Error("Could not get token from tokenGetter function.");
            }
            if (this.skipWhenExpired) {
                tokenIsExpired = token ? this.jwtHelper.isTokenExpired(token) : true;
            }
            if (token && tokenIsExpired && this.skipWhenExpired) {
                request = request.clone();
            }
            else if (token) {
                request = request.clone({
                    setHeaders: (_a = {},
                        _a[this.headerName] = "" + authScheme + token,
                        _a),
                });
            }
            return next.handle(request);
        };
        JwtInterceptor.prototype.intercept = function (request, next) {
            var _this = this;
            if (!this.isAllowedDomain(request) || this.isDisallowedRoute(request)) {
                return next.handle(request);
            }
            var token = this.tokenGetter(request);
            if (token instanceof Promise) {
                return rxjs.from(token).pipe(operators.mergeMap(function (asyncToken) {
                    return _this.handleInterception(asyncToken, request, next);
                }));
            }
            else {
                return this.handleInterception(token, request, next);
            }
        };
        return JwtInterceptor;
    }());
    JwtInterceptor.decorators = [
        { type: core.Injectable }
    ];
    JwtInterceptor.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [JWT_OPTIONS,] }] },
        { type: JwtHelperService },
        { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
    ]; };

    var JwtModule = /** @class */ (function () {
        function JwtModule(parentModule) {
            if (parentModule) {
                throw new Error("JwtModule is already loaded. It should only be imported in your application's main module.");
            }
        }
        JwtModule.forRoot = function (options) {
            return {
                ngModule: JwtModule,
                providers: [
                    {
                        provide: http.HTTP_INTERCEPTORS,
                        useClass: JwtInterceptor,
                        multi: true,
                    },
                    options.jwtOptionsProvider || {
                        provide: JWT_OPTIONS,
                        useValue: options.config,
                    },
                    JwtHelperService,
                ],
            };
        };
        return JwtModule;
    }());
    JwtModule.decorators = [
        { type: core.NgModule }
    ];
    JwtModule.ctorParameters = function () { return [
        { type: JwtModule, decorators: [{ type: core.Optional }, { type: core.SkipSelf }] }
    ]; };

    /*
     * Public API Surface of nativescript-angular-jwt
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.JWT_OPTIONS = JWT_OPTIONS;
    exports.JwtHelperService = JwtHelperService;
    exports.JwtInterceptor = JwtInterceptor;
    exports.JwtModule = JwtModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
//# sourceMappingURL=nativescript-angular-jwt.umd.js.map
